### PR TITLE
dialects: (hw) Add hw.instance operation

### PR DESCRIPTION
**Pull request by Moxinilian**
This PR adds a subset of the hw.instance operation, the dual of hw.module for instantiation.

### CODE REVIEW

**Comment by Moxinilian** [comment]
Codecov you have no idea how happy I am right now.

**Comment by PapyChacal** [comment]
Can you add an example with extra attributes on an `hw.instance`? I think the printer is missing it?

**Comment by Moxinilian** [comment]
Nice catch

**Review by lucjaulmes — APPROVED** [review]
Some tiny nits but otherwise LGTM.

File: xdsl/dialects/hw.py [hunk 1]
```diff
@@ -816,6 +821,207 @@ def print_port(port: ModulePort):
...
+            result_args,
+            (result.data for result in self.result_names.data),
+            "output port",
+        )
+
+    @classmethod
+    def parse(cls, parser: Parser) -> "InstanceOp":
+        instance_name = parser.parse_str_literal("instance name")
+        if parser.parse_optional_keyword("sym") is not None:
+            parser.raise_error("Instance inner symbols are not supported yet")
```
[the remark this answers is not in the document]
Comment (Moxinilian) [reply xdsl/dialects/hw.py:923]: Honestly I just followed advice I got from an annoyed Michel that I should keep new features to the bare minimum, so I did just that. You're right it's probably fairly simple, I did not even give it any thought (for instance parameters they can't even be reasonably parsed in xDSL right now).

[the remark this answers is not in the document]
Comment (Moxinilian) [reply xdsl/dialects/hw.py:923]: To be honest after trying it out in circt-opt inner symbols on instances have a lot of restrictions for them to be supported properly, it's not just about parsing them. Are you sure this is desired? I don't really know what inner symbols are about and have never encountered them so far.

[the remark this answers is not in the document]
Comment (Moxinilian) [reply xdsl/dialects/hw.py:923]: I have a mediocre implementation done that accept way more inner symbols than upstream CIRCT does. Tell me if that is good to you or if I should revert it.

[the remark this answers is not in the document]
Comment (lucjaulmes) [reply xdsl/dialects/hw.py]: > instances have a lot of restrictions for them to be supported properly, it's not just about parsing them. Are you sure this is desired?

I’m not – it really was just an open question. A laxer acceptation of inner symbols or leaving implementation out of this PR are both fine.

**Pull request merged**